pkg/registrar/map_functions_test.go

```diff
@@ -1248,16 +1278,205 @@ var _ = Describe("MapFunctions", func() {
 					It("returns an empty request list", func() {
 						sa := &corev1.ServiceAccount{
 							ObjectMeta: metav1.ObjectMeta{
-								Name:      "some-other-service-account",
-								Namespace: "some-other-namespace",
+								Name:      "some-unused-service-account",
+								Namespace: "some-unused-namespace",
 							},
 						}
 						reqs := m.ServiceAccountToWorkloadRequests(sa)
 
 						Expect(reqs).To(HaveLen(0))
 					})
 				})
+			})
+
+			Context("there is a matching workload from the supply chain", func() {
+				BeforeEach(func() {
+					existingWorkload := &v1alpha1.Workload{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "some-workload",
+							Namespace: "some-namespace",
+							Labels: map[string]string{
+								"some-label": "some-label-value",
+							},
+						},
+					}
+					existingWorkloadList := v1alpha1.WorkloadList{
+						Items: []v1alpha1.Workload{*existingWorkload},
+					}
+
+					existingSupplyChain := &v1alpha1.ClusterSupplyChain{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "some-supply-chain",
+						},
+						Spec: v1alpha1.SupplyChainSpec{
+							Selector: map[string]string{
+								"some-label": "some-label-value",
+							},
+							ServiceAccountName: "some-service-account",
+						},
+					}
+
+					existingSupplyChainList := v1alpha1.ClusterSupplyChainList{
+						Items: []v1alpha1.ClusterSupplyChain{*existingSupplyChain},
+					}
+
+					fakeClient.ListStub = func(ctx context.Context, list client.ObjectList, options ...client.ListOption) error {
+						listVal := reflect.Indirect(reflect.ValueOf(list))
+						switch list.(type) {
+						case *v1alpha1.WorkloadList:
+							existingVal := reflect.Indirect(reflect.ValueOf(existingWorkloadList))
+							listVal.Set(existingVal)
+						case *v1alpha1.ClusterSupplyChainList:
+							existingVal := reflect.Indirect(reflect.ValueOf(existingSupplyChainList))
+							listVal.Set(existingVal)
+						default:
+							panic("list type not stubbed")
+						}
+
+						return nil
+					}
+				})
+				It("returns requests for the matching workload", func() {
+					sa := &corev1.ServiceAccount{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "some-service-account",
+							Namespace: "some-namespace",
+						},
+					}
+					reqs := m.ServiceAccountToWorkloadRequests(sa)
+
+					Expect(reqs).To(HaveLen(1))
+					Expect(reqs[0].Name).To(Equal("some-workload"))
+				})
+			})
+
+			Context("a workload matches by itself and through a supply chain", func() {
+				BeforeEach(func() {
+					existingWorkload := &v1alpha1.Workload{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "some-workload",
+							Namespace: "some-namespace",
+							Labels: map[string]string{
+								"some-label": "some-label-value",
+							},
+						},
+						Spec: v1alpha1.WorkloadSpec{
+							ServiceAccountName: "some-service-account",
+						},
```

Comment:
There are 50 lines of boilerplate in the BeforeEach, but the only thing that should vary between the last test and this one is this. It's difficult to verify that the rest of the boilerplate is the same. That's why I advocate for deduplicating tests. Following how objects are built up from parent contexts is much easier than manually comparing dozens of lines of code for changes.

Reply:
I completely agree and I began to refactor the tests to do exactly that but after discussing with others we determined that we would like to refactor the whole map_functions.go and so it didn't seem worth it to refactor the tests right now if we will (hopefully) soon begin another refactor.